src/bit_machine/mod.rs

```diff
@@ -472,6 +506,37 @@ impl BitMachine {
     }
 }
 
+/// A struct that keeps track of which case branches were executed
+/// during the execution of the Bit Machine.
+///
+/// The trait is implemented for `[HashSet<Imr>; 2]`, which does the actual tracking,
+/// and it is implemented for `()`, which is a dummy tracker that is optimized out by the compiler.
+///
+/// The trait enables us to turn tracking on or off depending on a generic parameter.
+trait CaseTracker {
+    /// Track the execution of the left branch of the case node with the given `imr`.
+    fn track_left(&mut self, imr: Imr);
+
+    /// Track the execution of the right branch of the case node with the given `imr`.
+    fn track_right(&mut self, imr: Imr);
+}
+
+impl CaseTracker for [HashSet<Imr>; 2] {
+    fn track_left(&mut self, imr: Imr) {
+        self[0].insert(imr);
+    }
+
+    fn track_right(&mut self, imr: Imr) {
+        self[1].insert(imr);
+    }
+}
+
+impl CaseTracker for () {
```

Comment:
In b90069e5b7c4be7c93a12e57c494a11aa45c91da:

We can do this in a followup PR, but we really ought to create a dedicated dummy type for this rather than using `()`.

It's not so important for a private trait but it's cleaner.